File: app/src/main/java/com/jakebarnby/filemanager/managers/BillingManager.java
package com.jakebarnby.filemanager.managers;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;

import com.android.billingclient.api.AcknowledgePurchaseParams;
import com.android.billingclient.api.AcknowledgePurchaseResponseListener;
import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.SkuDetails;
import com.android.billingclient.api.SkuDetailsParams;
import com.android.billingclient.api.SkuDetailsResponseListener;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.jakebarnby.filemanager.util.Constants;
import com.jakebarnby.filemanager.util.LogUtils;
import com.jakebarnby.filemanager.util.PreferenceUtils;
import com.jakebarnby.filemanager.util.Utils;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Jake on 10/21/2017.
 */

public class BillingManager {

    private BillingClient mBillingClient;
    private int mRetryCount;


    public BillingManager(Context context) {
        mRetryCount = 0;

        mBillingClient = BillingClient.newBuilder(context).setListener((billingResult, purchases) -> {
            if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK && purchases != null) {
                for (Purchase purchase : purchases) {
                    handlePurchase(context, purchase);
                }
            } else if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.USER_CANCELED) {
                LogUtils.logFirebaseEvent(
                    FirebaseAnalytics.getInstance(context),
                    Constants.Analytics.EVENT_CANCELLED_PURCHASE);
            }
        }).enablePendingPurchases().build();

        mBillingClient.startConnection(new BillingClientStateListener() {
            @Override
            public void onBillingSetupFinished(BillingResult billingResult) {
                if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                    queryPurchases(context);
                }
            }

            @Override
            public void onBillingServiceDisconnected() {
                if (mRetryCount < 5) {
                    mBillingClient.startConnection(this);
                    mRetryCount++;
                }
            }
        });
    }

    /**
     * Start google play billing flow for the given sku
     *
     * @param activity Activity that launched the flow
     * @param sku      Sku of the product to purchase
     */
    public void purchaseItem(Activity activity, String sku) {
        List<String> skuList = new ArrayList<>();
        skuList.add(sku);

        SkuDetailsParams params = SkuDetailsParams
            .newBuilder()
            .setSkusList(skuList)
            .setType(BillingClient.SkuType.INAPP)
            .build();

        mBillingClient.querySkuDetailsAsync(params,
            (billingResult, skuDetailsList) -> {
                if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                    for (SkuDetails detail : skuDetailsList) {
                        if (detail.getSku().equals(sku)) {
                            BillingFlowParams flowParams = BillingFlowParams
                                .newBuilder()
                                .setSkuDetails(detail)
                                .build();

                            mBillingClient.launchBillingFlow(activity, flowParams);
                        }
                    }
                }
            });
    }

    /**
     * Handle a purchase success event
     *
     * @param context  Context for resources
     * @param purchase The purchase that was made
     */
    private void handlePurchase(Context context, Purchase purchase) {
        if (purchase.getPurchaseState() != Purchase.PurchaseState.PURCHASED) {
            // TODO: LOG error
            return;
        }

        switch (purchase.getSku()) {
            case Constants.Billing.SKU_PREMIUM:
                //TODO: Show dialog "Thanks! All ads have been disabled."
                PreferenceUtils.savePref(context, Constants.Prefs.HIDE_ADS_KEY, true);
                break;
        }

        if (!purchase.isAcknowledged()) {
            AcknowledgePurchaseParams acknowledgePurchaseParams =
                AcknowledgePurchaseParams.newBuilder()
                    .setPurchaseToken(purchase.getPurchaseToken())
                    .build();
            mBillingClient.acknowledgePurchase(
                acknowledgePurchaseParams,
                billingResult -> {
                    if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                        Bundle params = new Bundle();
                        params.putString(Constants.Analytics.PARAM_PURCHASE_SKU, purchase.getSku());
                        LogUtils.logFirebaseEvent(
                            FirebaseAnalytics.getInstance(context),
                            Constants.Analytics.EVENT_SUCCESS_PURCHASE,
                            params);
                    }
                }
            );
        }
    }

    /**
     * Check a users current purchases
     *
     * @param context Context for resources
     */
    private void queryPurchases(Context context) {
        Purchase.PurchasesResult purchasesResult = mBillingClient.queryPurchases(BillingClient.SkuType.INAPP);
        List<Purchase> purchases = purchasesResult.getPurchasesList();
        for (Purchase purchase : purchases) {
            if (purchase.getSku().equals(Constants.Billing.SKU_PREMIUM)) {
                PreferenceUtils.savePref(context, Constants.Prefs.HIDE_ADS_KEY, true);
            }
        }
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/managers/BillingManager.kt
package com.jakebarnby.filemanager.managers

import android.app.Activity
import android.content.Context
import androidx.core.os.bundleOf
import com.android.billingclient.api.*
import com.android.billingclient.api.BillingClient.BillingResponseCode.OK
import com.android.billingclient.api.BillingClient.BillingResponseCode.USER_CANCELED
import com.google.firebase.analytics.FirebaseAnalytics
import com.jakebarnby.filemanager.util.Constants
import com.jakebarnby.filemanager.util.Constants.Billing
import com.jakebarnby.filemanager.util.Constants.Prefs
import com.jakebarnby.filemanager.util.Logger
import kotlinx.coroutines.*

/**
 * Created by Jake on 10/21/2017.
 */
class BillingManager(
    context: Context,
    private val prefs: PreferenceManager
) {

    companion object {
        private const val MAX_RETRIES = 5
    }

    private val billingClient: BillingClient
    private var retryCount = 0

    init {
        billingClient = BillingClient.newBuilder(context).setListener { billingResult, purchases ->
            if (billingResult.responseCode == OK && purchases != null) {
                GlobalScope.launch {
                    purchases.map {
                        GlobalScope.async {
                            handlePurchase(context, it)
                        }
                    }.awaitAll()
                }

            } else if (billingResult.responseCode == USER_CANCELED) {
                // TODO: show snackbar purchase cancelled
                Logger.logFirebaseEvent(
                    FirebaseAnalytics.getInstance(context),
                    Constants.Analytics.EVENT_CANCELLED_PURCHASE)
            }
        }.enablePendingPurchases().build()

        billingClient.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(billingResult: BillingResult) {
                if (billingResult.responseCode == OK) {
                    queryPurchases(context)
                }
            }

            override fun onBillingServiceDisconnected() {
                if (retryCount < MAX_RETRIES) {
                    billingClient.startConnection(this)
                    retryCount++
                }
            }
        })
    }

    /**
     * Start google play billing flow for the given sku
     *
     * @param activity Activity that launched the flow
     * @param sku      Sku of the product to purchase
     */
    suspend fun purchaseItem(activity: Activity?, sku: String?) {
        val params = SkuDetailsParams.newBuilder()
            .setSkusList(listOf(sku))
            .setType(BillingClient.SkuType.INAPP)
            .build()

        val skuDetailResult = withContext(Dispatchers.IO) {
            billingClient.querySkuDetails(params)
        }
        val success = skuDetailResult.billingResult.responseCode == OK
        val skuDetailList = skuDetailResult.skuDetailsList

        if (!success || skuDetailList == null) {
            // TODO: Show failed to get sku error
            return
        }

        for (detail in skuDetailResult.skuDetailsList!!) {
            if (detail.sku != sku) {
                continue
            }
            val flowParams = BillingFlowParams.newBuilder()
                .setSkuDetails(detail)
                .build()
            billingClient.launchBillingFlow(activity, flowParams)
        }
    }

    /**
     * Handle a purchase success event
     *
     * @param context  Context for resources
     * @param purchase The purchase that was made
     */
    private suspend fun handlePurchase(context: Context, purchase: Purchase) {
        if (purchase.purchaseState != Purchase.PurchaseState.PURCHASED) {
            // TODO: LOG error
            return
        }

        when (purchase.sku) {
            Billing.SKU_PREMIUM -> {
                //TODO: Show dialog "Thanks! All ads have been disabled."
                prefs.savePref(Prefs.HIDE_ADS_KEY, true)
            }
        }

        if (purchase.isAcknowledged) {
            return
        }

        val params = AcknowledgePurchaseParams.newBuilder()
            .setPurchaseToken(purchase.purchaseToken)
            .build()

        val acknowledgeResult = withContext(Dispatchers.IO) {
            billingClient.acknowledgePurchase(params)
        }

        if (acknowledgeResult.responseCode == OK) {
            val args = bundleOf(
                Constants.Analytics.PARAM_PURCHASE_SKU to purchase.sku
            )
            Logger.logFirebaseEvent(
                FirebaseAnalytics.getInstance(context),
                Constants.Analytics.EVENT_SUCCESS_PURCHASE,
                args)
        }
    }

    /**
     * Check a users current purchases
     *
     * @param context Context for resources
     */
    private fun queryPurchases(context: Context) {
        val purchasesResult = billingClient.queryPurchases(BillingClient.SkuType.INAPP)
        val purchases = purchasesResult.purchasesList
        for (purchase in purchases) {
            if (purchase.sku == Billing.SKU_PREMIUM) {
                prefs.savePref(Prefs.HIDE_ADS_KEY, true)
            }
        }
    }
}